machine_learning/automatic_differentiation.py

```diff
@@ -0,0 +1,324 @@
+"""
+Demonstration of the Automatic Differentiation (Reverse mode).
+
+Reference: https://en.wikipedia.org/wiki/Automatic_differentiation
+
+Author: Poojan smart
+Email: smrtpoojan@gmail.com
+
+Examples:
+
+>>> with GradientTracker() as tracker:
+...     a = Variable([2.0, 5.0])
+...     b = Variable([1.0, 2.0])
+...     m = Variable([1.0, 2.0])
+...     c = a + b
+...     d = a * b
+...     e = c / d
+>>> print(tracker.gradient(e, a))
+[-0.25 -0.04]
+>>> print(tracker.gradient(e, b))
+[-1.   -0.25]
+>>> print(tracker.gradient(e, m))
+None
+
+>>> with GradientTracker() as tracker:
+...     a = Variable([[2.0, 5.0]])
+...     b = Variable([[1.0], [2.0]])
+...     c = a @ b
+>>> print(tracker.gradient(c, a))
+[[1. 2.]]
+>>> print(tracker.gradient(c, b))
+[[2.]
+ [5.]]
+
+>>> with GradientTracker() as tracker:
+...     a = Variable([[2.0, 5.0]])
+...     b = a ** 3
+>>> print(tracker.gradient(b, a))
+[[12. 75.]]
+"""
+from __future__ import annotations
+
+from enum import Enum
+from types import TracebackType
+
+import numpy as np
+from typing_extensions import Self  # noqa: UP035
+
+
+class Variable:
+    """
+    Class represents n-dimensional object which is used to wrap
+    numpy array on which operations will be performed and gradient
+    will be calculated.
+    """
+
+    def __init__(self, value) -> None:
```

Comment:
`from typing import Any`
```suggestion
    def __init__(self, value: Any) -> None:
```